docs/root/intro/arch_overview/data_sharing_between_filters.rst

```diff
@@ -0,0 +1,93 @@
+.. _arch_overview_data_sharing_between_filters:
+
+Envoy provides the following mechanisms for the transfer of configuration,
+metadata and per-request/connection state to, from and between filters, as
+well as to other core subsystems  (e.g., access logging).
+
+
+Static State
+============
+
+Static state is any immutable state specified at configuration load time
+(e.g., through XDS). There are three categories of static state:
+
+Metadata
+--------
+
+Several parts of Envoy configuration (e.g. listeners, routes, clusters )
+contain a metadata field (map<string, Protobuf::Struct>) where arbitrary
+key-value pairs can be encoded.  The typical pattern is to use the filter
+names in reverse DNS format as the key and encode filter specific
+configuration metadata in the value. This metadata is immutable and shared
+across all requests/connections. Such config metadata is usually provided
+during bootstrap time or as part of XDS. For example, weighted clusters in
+HTTP routes use the metadata to indicate the labels on the endpoints
+corresponding to the weighted cluster. Another example, the subset load
+balancer uses the metadata from the route entry corresponding to the
+weighted cluster to select appropriate endpoints in a cluster
+
+Typed Metadata
+--------------
+
+Metadata as such is untyped. It incurs a performance const when the caller
+wants to convert it to some typed class object and act on it. Typed
+Metadata solves this problem by allowing Filters to register a one time
+conversion logic for a specific key. Incoming config metadata (via XDS) is
+converted to class objects at config load time. Filters can then obtain a
+typed variant of the metadata (usually protobuf) at runtime (per
+request/connection), thereby eliminating the need for filters to repeatedly
+convert from Protobuf::Struct to some internal object during
+request/connection processing.
+
+For example, if a filter that desires to have a convenience wrapper class
+over the opaque metadata with key "xxx.service.policy" in ClusterInfo, will
+register a ServicePolicyFactory that inherits from
+ClusterTypedMetadataFactory. The factory translates the protobuf.Struct
+into an instance of ServicePolicy class (inherited from
+FilterState::Object). When a Cluster is created, the associated
+ServicePolicy instance will be created and cached.
+
```

Comment:
Probably helpful to the reader to re-emphasize here that typed metadata is not a new metadata source, and that the underlying untyped metadata is canonical.